feat(jwt): [#4191] extend Token.encode() to support custom headers

s-aleshin opened the pull request:
<!--
By submitting this pull request, you agree to:
- follow [Litestar's Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)
- follow [Litestar's contribution guidelines](https://github.com/litestar-org/.github/blob/main/CONTRIBUTING.md)
- follow the [PSFs's Code of Conduct](https://www.python.org/psf/conduct/)
-->
## Description
Allows passing custom JWT headers (e.g., kid) to Token.encode(). This improves flexibility in cases where custom JWT headers (like `kid`) are required.
-

<!--
Please add in issue numbers this pull request will close, if applicable
Examples: Fixes #4321 or Closes #1234

Ensure you are using a supported keyword to properly link an issue:
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
## Closes
Fixes #4191

provinzkraut reviewed (changes requested):
Two things:

1. This is a breaking change (for users that subclass `Token`), so it would have to be targeted at the `v3` branch
2. I don't think we should add `json_encoder` and `sort_headers`, as they are specific to `pyjwt`, and this is meant as an abstraction *over* `pyjwt`, not a drop-in replacement. If you do need to customise how `decode` interacts with `pyjwt`, you can simply subclass `Token`

s-aleshin commented:
> Two things:
> 
> 1. This is a breaking change (for users that subclass `Token`), so it would have to be targeted at the `v3` branch
> 2. I don't think we should add `json_encoder` and `sort_headers`, as they are specific to `pyjwt`, and this is meant as an abstraction _over_ `pyjwt`, not a drop-in replacement. If you do need to customise how `decode` interacts with `pyjwt`, you can simply subclass `Token`

1. I’ve rebased the branch onto v3.0 as you suggested — the PR now targets the correct base.
2. Ok, I agree. `json_encoder` and `sort_headers` are specific for `pyjwt` and use rarely. Deleted from the PR.

provinzkraut commented on litestar/security/jwt/token.py:
```diff
@@ -194,12 +195,18 @@ def decode(
-    def encode(self, secret: str, algorithm: str) -> str:
+    def encode(
+        self,
+        secret: AllowedPrivateKeys | str | bytes,
```
`str | bytes` is fine but we don't want to expose the pyjwt internals (`AllowedPrivateKeys`) here

provinzkraut reviewed (approved):
Thanks @s-aleshin!
